Feature: Add support for IPython in volshell

dgmcdona opened the pull request:
Hi! This adds support for the IPython interactive console in `volshell`. I've tested to make sure the `combined_locals` are correctly inserted into the user namespace, the banner displays correctly on startup, and the prompt includes the layer name.

dgmcdona committed:
Feature: Add support for IPython in volshell

ikelos commented on volatility3/cli/volshell/generic.py:
```diff
@@ -69,43 +78,66 @@ def run(
+                import readline
+            except ImportError:
+                pass
+            else:
```
This is a bit strange, and unintuitive to follow, should it no be:
```
try:
  import readline
  import rlcompleter
  completer = ...
except ImportError:
  # No readline support
  pass
```
instead?  Doesn't seem useful to have the exception handler buried in the middle of the stuff that happens if readline is imported?  Plus this also protected against `rlcompleter` not existing...

ikelos reviewed (commented):
Happy with the concept, just a bit of an odd structure in the middle of it...

dgmcdona committed:
Volshell: Fix import logic around readline + rlcomplete

dgmcdona committed:
Volshell: Handle script running in ipython shell

ikelos commented on volatility3/cli/volshell/generic.py:
```diff
@@ -69,43 +78,65 @@
+                readline.set_completer(completer.complete)
+                readline.parse_and_bind("tab: complete")
+                print("Readline imported successfully")
+            except ImportError:
```
Thanks, could you include at least a comment here please?

ikelos commented on volatility3/cli/volshell/generic.py:
```diff
@@ -508,10 +539,11 @@
-            self.__console.runsource(
-                io.TextIOWrapper(fp, encoding="utf-8").read(), symbol="exec"
-            )
+        with io.TextIOWrapper(accessor.open(url=location), encoding="utf-8") as fp:
```
Err, might be worth doing the accessor.open and then wrapping it in `io.TextIOWrapper` inside the context manager?

ikelos reviewed (commented):
Closer, just a couple extra bits now please...

dgmcdona committed:
Volshell: Address comments from code review

- Add failure message when readline or rlcompleter can't be imported
- Fix unclosed file handle in context manager

dgmcdona commented:
Thanks!
